python/ml4ir/data/tfrecord_reader.py

```diff
@@ -68,76 +65,117 @@ def _parse_sequence_example_fn(sequence_example_proto):
+                            target_width=max_num_records,
+                        )
+
+                    mask = tf.cond(
+                        tf.shape(mask)[1] < max_num_records,
+                        # Pad if there are missing records
+                        lambda: image.pad_to_bounding_box(
+                            mask,
+                            offset_height=0,
+                            offset_width=0,
+                            target_height=1,
+                            target_width=max_num_records,
+                        ),
+                        # Crop if there are extra records
+                        crop_fn,
                     )
-                    features_dict["mask"] = tf.squeeze(mask)
+                    mask = tf.squeeze(mask)
 
-                t = sparse.reset_shape(t, new_shape=[1, max_num_records])
-                t = sparse.to_dense(t)
-                t = tf.squeeze(t)
+                    # Check validity of mask
+                    tf.debugging.assert_greater(
+                        tf.cast(tf.reduce_sum(mask), tf.float32), tf.constant(0.0)
+                    )
+
+                    features_dict["mask"] = mask
+
+                feature_tensor = sparse.reset_shape(feature_tensor, new_shape=[1, max_num_records])
+                feature_tensor = sparse.to_dense(feature_tensor)
+                feature_tensor = tf.squeeze(feature_tensor)
 
                 # If feature is a string, then decode into numbers
-                if feature_config.get_dict()[feat]["type"] == FeatureTypeKey.STRING:
-                    t = io.decode_raw(
-                        t,
+                if feature_layer_info["type"] == FeatureTypeKey.STRING:
+                    feature_tensor = io.decode_raw(
+                        feature_tensor,
                         out_type=tf.uint8,
-                        fixed_length=feature_config.get_dict()[feat]["max_length"],
+                        fixed_length=feature_layer_info["max_length"],
                     )
-                    t = tf.cast(t, tf.float32)
+                    feature_tensor = tf.cast(feature_tensor, tf.float32)
             else:
                 #
                 # Handle dense tensors
                 #
                 # if len(t.shape) == 1:
-                #     t = tf.expand_dims(t, axis=0)
+                #     feature_tensor = tf.expand_dims(t, axis=0)
                 # if len(t.shape) == 2:
-                #     t = tf.pad(t, paddings=[[0, 0], [0, max_num_records]])
-                #     t = tf.squeeze(t)
+                #     feature_tensor = tf.pad(t, paddings=[[0, 0], [0, max_num_records]])
+                #     feature_tensor = tf.squeeze(t)
                 # else:
                 #     raise Exception('Invalid input : {}'.format(feat))
-                raise ValueError("Invalid input : {}".format(feat))
+                raise ValueError("Invalid input : {}".format(feature_name))
```

Comment:
so we still don't handle dense?  MRUAffinity features aren't dense?  What does this else cover?

Reply:
It doesn't cover anything for now. Since, the way I define the feature spec for parsing the TFRecord is a bunch of `tf.io.VarLenFeature`, this should not be called at all for sequence features. I had it initially when I was actually getting dense features. But that was long ago. Had let the code be in case we were to get dense features in the future. But I don't think we will. Can remove.

Reply:
In order to get dense features here, I need to define each feature(say MRU Affinity) as a `FixedLenFeature`(additionally specifying that fixed length/shape), which they are not - since each query has different number of records. Hence I use `tf.io.ValLenFeature`

Reply:
Got it, yes VarLenFeature because of variable record counts